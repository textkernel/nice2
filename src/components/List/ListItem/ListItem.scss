$contexts: (neutral, brand, primary, accent, info, good, warning, bad);

.ListItem {
    transition: background-color var(--transition-duration);

    &__container {
        display: flex;
        align-items: baseline;
        padding: var(--spacing-4x);
    }

    &:hover:not(.ListItem--disabled) {
        background-color: var(--color-highlight-background);
    }

    &--isSelected, &--isHighlighted {
        background-color: var(--color-highlight-background);
    }

    &--clickable {
        cursor: pointer;
    }

    &--highlightContext {
        @each $context in $contexts {
            &_#{ $context }:hover:not(.ListItem--disabled) {
                background-color: var(--color-#{$context});
                color: var(--color-background);
            }

            &_#{ $context }.ListItem--isSelected {
                background-color: var(--color-#{$context});
                color: var(--color-background);
            }

            &_#{ $context }.ListItem--isHighlighted:not(.ListItem--disabled) {
                background-color: var(--color-#{$context});
                color: var(--color-highlight-background)
            }
        }
    }
}
