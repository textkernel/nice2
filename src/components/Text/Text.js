import React from 'react';
import PropTypes from 'prop-types';
import bem from 'bem';
import styles from './Text.scss';
import { SIZES, CONTEXTS } from '../../constants';

const { block } = bem({
    name: 'Text',
    classnames: styles,
    propsToMods: ['context', 'size'],
});

const Text = props => {
    const { children, context, inline, size, ...rest } = props;
    const HtmlNodeType = inline ? 'span' : 'p';

    return (
        <HtmlNodeType {...rest} {...block(props)}>
            {children}
        </HtmlNodeType>
    );
};

Text.displayName = 'Text';

Text.propTypes = {
    /** Text content */
    children: PropTypes.oneOfType([PropTypes.string, PropTypes.node]).isRequired,
    /** Text should be rendered inline */
    inline: PropTypes.bool,
    /** The context of the text, effecting its color (e.g. brand, primary, bad, good etc. 'muted' added as special context here) */
    context: PropTypes.oneOf([...CONTEXTS, 'muted', 'default']),
    /** Custom text sizes */
    size: PropTypes.oneOf(SIZES),
};

Text.defaultProps = {
    inline: false,
    context: 'default',
    size: 'normal',
};

export default Text;
