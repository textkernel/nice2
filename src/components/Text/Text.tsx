import * as React from 'react';
import bem from '../../utils/bem';
import styles from './Text.scss';
import { Size, Context } from '../../constants';

interface Props extends React.HTMLAttributes<HTMLElement> {
    /** Text content */
    children: React.ReactElement;
    /** Text should be rendered inline */
    inline: boolean;
    /** The context of the text, effecting its color (e.g. brand, primary, bad, good etc. 'muted' added as special context here) */
    context: Context | 'muted' | 'default';
    /** Custom text sizes */
    size: Size;
}

const { block } = bem('Text', styles);

const Text: React.FC<Props> = props => {
    const { children, context, inline, size, ...rest } = props;
    const HtmlNodeType = inline ? 'span' : 'p';

    return (
        <HtmlNodeType {...rest} {...block(props)}>
            {children}
        </HtmlNodeType>
    );
};

Text.displayName = 'Text';

Text.defaultProps = {
    inline: false,
    context: 'default',
    size: 'normal',
};

export default Text;
