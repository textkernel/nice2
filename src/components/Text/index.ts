export { Text } from './Text';
export { MarkedText } from './MarkedText';
