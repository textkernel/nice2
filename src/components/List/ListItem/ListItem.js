import * as React from 'react';
import PropTypes from 'prop-types';
import { bem } from '../../../utils/bem';
import { Text } from '../../Text';
import { CONTEXTS } from '../../../constants';
import styles from './ListItem.scss';

const { block, elem } = bem('ListItem', styles);

export const ListItem = React.forwardRef((props, ref) => {
    const {
        children,
        isSelected,
        isHighlighted,
        onClick,
        disabled,
        highlightContext,
        ...rest
    } = props;
    const customBlockMod = { clickable: typeof onClick === 'function' };

    return (
        <li {...rest} ref={ref} {...block({ ...props, ...customBlockMod })}>
            <div onClick={onClick} role="presentation" {...elem('container', props)}>
                {React.Children.map(children, child =>
                    typeof child === 'string' ? <Text inline>{child}</Text> : child
                )}
            </div>
        </li>
    );
});

ListItem.displayName = 'ListItem';

ListItem.propTypes = {
    /** List Items */
    children: PropTypes.node,
    /** A function to be called if the item is clicked */
    onClick: PropTypes.func,
    /** Formats this item as selected  */
    isSelected: PropTypes.bool,
    /** Formats this item as selected  */
    isHighlighted: PropTypes.bool,
    /** Format this item as disabled */
    disabled: PropTypes.bool,
    /** formatting context when hovered or selected */
    highlightContext: PropTypes.oneOf([...CONTEXTS, 'default']),
};

ListItem.defaultProps = {
    children: null,
    onClick: null,
    isSelected: false,
    isHighlighted: false,
    disabled: false,
    highlightContext: 'default',
};
