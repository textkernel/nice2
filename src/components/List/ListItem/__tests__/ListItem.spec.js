import React from 'react';
import toJson from 'enzyme-to-json';
import ListItem from '../ListItem';
import Text from '../../../Text';

describe('ListItem component', () => {
    it('should render ListItem correctly', () => {
        const ref = React.createRef();
        const wrapper = mount(
            <ListItem ref={ref}>
                <Text>An item</Text>
            </ListItem>
        );

        expect(toJson(wrapper)).toMatchSnapshot();
        expect(ref.current.props.Component.displayName).toEqual('ListItem');
    });

    it('should turn string items to Text', () => {
        const wrapper = mount(<ListItem>Simple string</ListItem>);

        expect(wrapper.find('Text')).toHaveLength(1);
    });
    it('should not add clickable class when onClick is not defined', () => {
        const wrapper = mount(<ListItem>An item</ListItem>);

        expect(wrapper.find('.ListItem--clickable')).toHaveLength(0);
    });
    it('should add clickable class when onClick is defined', () => {
        const wrapper = mount(<ListItem onClick={jest.fn()}>An item</ListItem>);

        expect(wrapper.find('.ListItem--clickable')).toHaveLength(1);
    });

    it('should call onClick function when clicked', () => {
        const onClick = jest.fn();
        const wrapper = mount(<ListItem onClick={onClick}>An item</ListItem>);
        wrapper.find('div').simulate('click');

        expect(onClick).toHaveBeenCalledTimes(1);
    });
});
