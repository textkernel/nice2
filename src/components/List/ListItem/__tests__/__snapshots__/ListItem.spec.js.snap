// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem component should render ListItem correctly 1`] = `
<ListItem
  disabled={false}
  isSelected={false}
  onClick={null}
>
  <li
    className="ListItem"
  >
    <div
      className="ListItem__container"
      onClick={null}
      role="presentation"
    >
      <Text
        context="default"
        inline={false}
        key=".0"
        size="normal"
      >
        <p
          className="Text"
        >
          An item
        </p>
      </Text>
    </div>
  </li>
</ListItem>
`;
